src/main/java/io/sqreen/sasdk/backend/IngestionHttpClient.java

```diff
@@ -144,69 +145,96 @@
         void reportTrace(Object trace) throws IOException;
     }
 
-    static class IngestionHttpClientImpl
-            implements WithAuthentication, WithoutAuthentication {
 
-        private final String host;
-        private final BackendHttpImpl backendHttp;
-        private final AuthenticationConfig config;
+    static class IngestionHttpAuthClientImpl extends IngestionHttpClientImpl  implements WithAuthentication {
 
-        public IngestionHttpClientImpl(String host,
-                                       BackendHttpImpl backendHttp,
-                                       AuthenticationConfig config) {
-            checkNotNull(host);
-            checkNotNull(backendHttp);
-            this.host = host;
-            this.backendHttp = backendHttp;
+        protected final AuthenticationConfig config;
+        private Multimap<String, String> headers = ArrayListMultimap.create();
+
+        public IngestionHttpAuthClientImpl(String host,
+                                           BackendHttpImpl backendHttp,
+                                           AuthenticationConfig config) {
+            super(host, backendHttp);
             this.config = config != null ? config : AuthenticationConfig.EmptyConfig.INSTANCE;
+
+            Optional<String> sessionKey = this.config.getSessionKey();
+            if (sessionKey.isPresent()) {
+                this.headers.put("X-Session-Key", sessionKey.get());
+            } else {
+                Optional<String> apiKey = this.config.getAPIKey();
+                if (apiKey.isPresent()) {
+                    this.headers.put("X-API-Key", apiKey.get());
+                    Optional<String> appName = this.config.getAppName();
+                    if (appName.isPresent()) {
+                        this.headers.put("X-App-Name", appName.get());
+                    }
+                }
+            }
         }
 
         @Override
-        public void reportSignal(Object signal, Multimap<String, String> headers) throws IOException {
-            doIngestionRequestNoAuth("signals", signal, headers);
+        public void reportBatch(Collection<?> signalsAndTraces) throws IOException {
+            super.reportBatch(signalsAndTraces, headers);
```

Comment:
`super` is redundant, because this method also exists in this class. With the way you wrote this, you still ended up with a class implementing both the authenticated and unauthenticated methods like before, except now with the methods strewn across a class hierarchy.

If you want to separate the two implementations (so that just the authenticated one has the headers and the auth config), IMO you shouldn't have used inheritance and should have just wrapped the unauthenticated client inside the authenticated one (i.e. use composition, not inheritance).